controllers/topology/reconcile_state_test.go

```diff
@@ -0,0 +1,361 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topology
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestReconcileCluster(t *testing.T) {
+	cluster1 := newFakeCluster(metav1.NamespaceDefault, "cluster1").Obj()
+	cluster1WithReferences := newFakeCluster(metav1.NamespaceDefault, "cluster1").
+		WithInfrastructureCluster(newFakeInfrastructureCluster(metav1.NamespaceDefault, "infrastructure-cluster1").Obj()).
+		WithControlPlane(newFakeControlPlane(metav1.NamespaceDefault, "control-plane1").Obj()).
+		Obj()
+	cluster2WithReferences := cluster1WithReferences.DeepCopy()
+	cluster2WithReferences.SetGroupVersionKind(cluster1WithReferences.GroupVersionKind())
+	cluster2WithReferences.Name = "cluster2"
+
+	tests := []struct {
+		name    string
+		current *clusterv1.Cluster
+		desired *clusterv1.Cluster
+		want    *clusterv1.Cluster
+		wantErr bool
+	}{
+		{
+			name:    "Should update the cluster if infrastructure and control plane references are not set",
+			current: cluster1,
+			desired: cluster1WithReferences,
+			want:    cluster1WithReferences,
+			wantErr: false,
+		},
+		{
+			name:    "Should be a no op if infrastructure and control plane references are already set",
+			current: cluster2WithReferences,
+			desired: cluster2WithReferences,
+			want:    cluster2WithReferences,
```

Comment:
```suggestion
			current: cluster1WithReferences,
			desired: cluster1WithReferences,
			want:    cluster1WithReferences,
```
Mostly to confirm my understanding of the code. This would also work right? Or is there something more behind using cluster2WithReferences which I'm missing? (currently I would assume it's mostly to keep the test cases clearly separated)

Reply:
(if it also works, we of course could then drop `cluster2WithReferences`, but up to you)

Reply:
This would also work, but it also imply a that two test cases overlaps - uses the same objects - which is something that is better to avoid (e.g this will create problem if for any reason we switch to testenv)

Reply:
Thx for the explanation :)